File: src/types/api/aiSearch.ts
export type CfsAiDomain =
  | "data_readiness"
  | "economics"
  | "flood"
  | "general"
  | "methodology"
  | "model_lab"
  | "permits"
  | "schools"
  | "transportation"
  | "utilities"
  | "zoning";

export interface CfsAiSearchRequest {
  app_mode?: "economics" | "master-data" | "planning";
  conversation_context?: CfsAiConversationTurn[];
  filter_context?: Record<string, string | number | boolean | null | undefined>;
  filters?: {
    domains?: CfsAiDomain[];
    year_end?: number | null;
    year_start?: number | null;
  };
  mode?: "demo" | "live";
  query: string;
  request_type?: "powerbi_report_plan" | null;
  selected_signal?: CfsAiSelectedSignal | null;
}

export interface CfsAiSelectedSignal {
  domain: string;
  evidence?: string[];
  id: string;
  related_layers?: string[];
  status_band?: string | null;
  title: string;
}

export interface CfsAiConversationTurn {
  answer_summary?: string | null;
  dashboard_actions?: CfsAiDashboardActions;
  focused_domain?: string | null;
  query: string;
  related_layers?: string[];
}

export interface CfsAiEvidenceItem {
  as_of?: string | null;
  caveat?: string | null;
  confidence: "available" | "limited" | "not_available";
  detail: string;
  geography?: string | null;
  methodology?: string | null;
  source: string;
  source_type?: string | null;
  status?: string | null;
  title: string;
  unit?: string | null;
  value?: string | number | null;
}

export interface CfsAiDashboardActions {
  filter_watchlist?: {
    domain?: string | null;
    status?: string | null;
  } | null;
  focus_domain?:
    | "data_readiness"
    | "economics"
    | "flood"
    | "general"
    | "model_lab"
    | "permits"
    | "schools"
    | "transportation"
    | "utilities"
    | "zoning"
    | null;
  highlight_kpis?: string[];
  open_detail?: {
    type: "domain" | "kpi" | "watchlist";
    id: string;
  } | null;
  recommended_layers?: string[];
  sort_watchlist_by?: "data_gap" | "recent_activity" | "severity" | null;
  time_range?: {
    end_year?: number | null;
    start_year?: number | null;
  } | null;
}

export interface CfsAiPowerBiActions {
  action_type:
    | "add_to_canvas"
    | "build_chart"
    | "build_report"
    | "configure_builder"
    | "none"
    | "suggest_report";
  chart_builder_config?: {
    aggregation?: "average" | "count" | "sum";
    category_field?: string;
    caveat?: string;
    chart_type?: "bar" | "donut" | "line" | "matrix" | "pie" | "table";
    filter_field?: string;
    filter_value?: string;
    table_name: string;
    title?: string;
    value_field?: string;
  } | null;
  powerbi_build_steps?: string[];
  report_canvas_items?: Array<{
    aggregation?: string;
    category_field?: string;
    caveat?: string;
    filter_field?: string;
    filter_value?: string;
    page_name: string;
    powerbi_recipe: string;
    source_table: string;
    value_field?: string;
    visual_title: string;
    visual_type: "bar" | "donut" | "line" | "matrix" | "pie" | "table";
  }>;
  report_summary?: string;
  report_title?: string;
  selected_filters?: {
    data_confidence?: string;
    economic_segment?: string;
    geography_label?: string;
    opportunity_class?: string;
    scenario_name?: string;
    sewer_proxy_class?: string;
    utility_capacity_status?: string;
    utility_readiness_proxy_class?: string;
  };
  selected_tool?:
    | "chart_builder"
    | "decision_pack"
    | "planning_model"
    | "powerbi_export"
    | "report_canvas"
    | "scenario_model";
}

export interface CfsAiSearchResponse {
  answer: string;
  answer_mode?: "deterministic" | "provider_enhanced" | "safety";
  as_of: string | null;
  caveats: string[];
  context_freshness?: "cached_demo_extract" | "current_session" | "fallback_partial" | string | null;
  dashboard_actions: CfsAiDashboardActions;
  data_source?: "local_live_backend" | "portfolio_demo_extract" | string | null;
  data_mode: "demo" | "live";
  domains: CfsAiDomain[];
  evidence: CfsAiEvidenceItem[];
  executive_summary?: string | null;
  fallback_used?: boolean;
  filtered_context_summary?: string | null;
  interpretation?: string | null;
  key_findings?: string[];
  limitations?: string[];
  official_data_still_needed?: string[];
  prompt_version?: string;
  provider: "none" | "openai";
  provider_status?: string | null;
  provenance?: Record<string, unknown>;
  recommended_next_actions?: string[];
  request_id?: string;
  response_time_ms?: number;
  powerbi_actions?: CfsAiPowerBiActions | null;
  related_layers: string[];
  suggested_actions: string[];
  suggested_follow_up_questions?: string[];
  timings_ms?: Record<string, number>;
}

File: backend/app/schemas/ai_search.py
from typing import Any, Literal
from uuid import uuid4

from pydantic import BaseModel, Field

CfsAiDomain = Literal[
    "data_readiness",
    "economics",
    "flood",
    "general",
    "methodology",
    "model_lab",
    "permits",
    "schools",
    "transportation",
    "utilities",
    "zoning",
]

CfsAiProvider = Literal["none", "openai"]
CfsAiDashboardFocusDomain = Literal[
    "data_readiness",
    "economics",
    "flood",
    "general",
    "model_lab",
    "permits",
    "schools",
    "transportation",
    "utilities",
    "zoning",
]


class CfsAiSearchFilters(BaseModel):
    domains: list[CfsAiDomain] = Field(default_factory=list)
    year_end: int | None = None
    year_start: int | None = None


class CfsAiConversationTurn(BaseModel):
    answer_summary: str | None = Field(default=None, max_length=500)
    dashboard_actions: dict[str, Any] = Field(default_factory=dict)
    focused_domain: str | None = Field(default=None, max_length=50)
    query: str = Field(max_length=500)
    related_layers: list[str] = Field(default_factory=list, max_length=8)


class CfsAiSelectedSignal(BaseModel):
    domain: str = Field(max_length=80)
    evidence: list[str] = Field(default_factory=list, max_length=8)
    id: str = Field(max_length=120)
    related_layers: list[str] = Field(default_factory=list, max_length=8)
    status_band: str | None = Field(default=None, max_length=80)
    title: str = Field(max_length=200)


class CfsAiSearchRequest(BaseModel):
    app_mode: Literal["economics", "planning"] = "planning"
    conversation_context: list[CfsAiConversationTurn] = Field(default_factory=list, max_length=5)
    filter_context: dict[str, Any] = Field(default_factory=dict)
    filters: CfsAiSearchFilters = Field(default_factory=CfsAiSearchFilters)
    mode: Literal["demo", "live"] = "live"
    query: str = Field(min_length=1, max_length=500)
    request_type: Literal["powerbi_report_plan"] | None = None
    selected_signal: CfsAiSelectedSignal | None = None


class CfsAiEvidenceItem(BaseModel):
    as_of: str | None = None
    caveat: str | None = None
    confidence: Literal["available", "limited", "not_available"] = "available"
    detail: str
    geography: str | None = None
    methodology: str | None = None
    source: str
    source_type: str | None = None
    status: str | None = None
    title: str
    unit: str | None = None
    value: str | float | int | None = None


class CfsAiWatchlistFilter(BaseModel):
    domain: str | None = None
    status: str | None = None


class CfsAiOpenDetailAction(BaseModel):
    type: Literal["domain", "kpi", "watchlist"]
    id: str


class CfsAiTimeRangeAction(BaseModel):
    end_year: int | None = None
    start_year: int | None = None


class CfsAiDashboardActions(BaseModel):
    filter_watchlist: CfsAiWatchlistFilter | None = None
    focus_domain: CfsAiDashboardFocusDomain | None = None
    highlight_kpis: list[str] = Field(default_factory=list)
    open_detail: CfsAiOpenDetailAction | None = None
    recommended_layers: list[str] = Field(default_factory=list)
    sort_watchlist_by: Literal["data_gap", "recent_activity", "severity"] | None = None
    time_range: CfsAiTimeRangeAction | None = None


class CfsAiSearchResponse(BaseModel):
    answer: str
    answer_mode: Literal["deterministic", "provider_enhanced", "safety"] = "deterministic"
    as_of: str | None = None
    caveats: list[str] = Field(default_factory=list)
    context_freshness: str | None = None
    dashboard_actions: CfsAiDashboardActions = Field(
        default_factory=CfsAiDashboardActions,
    )
    data_source: str | None = None
    data_mode: Literal["demo", "live"] = "live"
    domains: list[CfsAiDomain] = Field(default_factory=list)
    evidence: list[CfsAiEvidenceItem] = Field(default_factory=list)
    executive_summary: str | None = None
    fallback_used: bool = False
    filtered_context_summary: str | None = None
    interpretation: str | None = None
    key_findings: list[str] = Field(default_factory=list)
    limitations: list[str] = Field(default_factory=list)
    official_data_still_needed: list[str] = Field(default_factory=list)
    powerbi_actions: dict[str, Any] | None = None
    prompt_version: str = "ask-cfs-2026-07-31"
    provider: CfsAiProvider = "none"
    provider_status: str | None = None
    provenance: dict[str, Any] = Field(default_factory=dict)
    recommended_next_actions: list[str] = Field(default_factory=list)
    related_layers: list[str] = Field(default_factory=list)
    request_id: str = Field(default_factory=lambda: str(uuid4()))
    response_time_ms: int = 0
    suggested_actions: list[str] = Field(default_factory=list)
    suggested_follow_up_questions: list[str] = Field(default_factory=list)
    timings_ms: dict[str, int] = Field(default_factory=dict)


CfsAiContext = dict[str, Any]
